tests/test_url.py

```diff
@@ -1104,6 +1104,18 @@ def test_scheme_case_insensitive(self):
         result = parse_data_uri("DaTa:,A%20brief%20note")
         self.assertEqual(result.data, b"A brief note")
 
+    def test_safe_url_string_encode_idna_domain_with_port(self):
+        self.assertEqual(safe_url_string('http://新华网.中国:80'),
+                         'http://xn--xkrr14bows.xn--fiqs8s:80')
+
+    def test_safe_url_string_encode_idna_domain_with_username_password_and_port_number(self):
+        self.assertEqual(safe_url_string('ftp://admin:admin@新华网.中国:21'),
+                         'ftp://admin:admin@xn--xkrr14bows.xn--fiqs8s:21')
+
+    def test_safe_url_string_encode_idna_domain_with_username_without_password_and_port_number(self):
+        self.assertEqual(safe_url_string('ftp://admin:@新华网.中国:21'),
+                         'ftp://admin:@xn--xkrr14bows.xn--fiqs8s:21')
```

Comment:
I believe this tests an empty password. No password would mean no colon: `…//admin@…`

Reply:
Did know about that, fixed and rename the test.